eoe/man/man1/gencat.1
'\"!  tbl | mmdoc
'\"macro stdmacro
.if n .pH g1.gencat @(#)gencat	41.10 of 5/26/91
.\" Copyright 1991 UNIX System Laboratories, Inc.
.\" Copyright 1989, 1990 AT&T
.nr X
.if \nX=0 .ds x} gencat 1 "" "\&"
.if \nX=1 .ds x} gencat 1 ""
.if \nX=2 .ds x} gencat 1 "" "\&"
.if \nX=3 .ds x} gencat "" "" "\&"
.TH \*(x}
.SH "NAME"
\f4gencat\f1 \- generate a formatted message catalogue
.SH "SYNOPSIS"
\f4gencat\f1
\f1[\f4\-m\f1]
.I catfile
.I msgfile
\ .\|.\|.
.SH "DESCRIPTION"
.br
The
\f4gencat\f1
utility
merges the message text source file(s) 
\f2msgfile\fP
into a formatted message database 
\f2catfile\f1.
The database
\f2catfile\f1
will be created if it does not already exist.
If 
\f2catfile\f1
does exist
its messages will be included in the new 
\f2catfile\f1.
If set and message numbers collide, 
the new message text defined in
\f2msgfile\f1
will replace the old message text currently contained in
\f2catfile\f1.
The message text source file (or set of files) input to
\f4gencat\f1
can contain either set and message numbers or simply
message numbers, in which case the set \f4NL_SETD\fP [see \f4nl_types\fP(5)]
is assumed.
.PP
The meanings of the options are:
.TP 6
\f4\-m\f1
Use memory allocated by the \f2malloc\f1() routine, instead of
memory mapped allocation routine \f2mmap\f1().
.TP
\f4\-t\f1 \f2textmax\f1
This is optional.  This parameter \f2textmax\f1 allows \f4gencat\f1 to
dynamically allocate the message text area.  If not specified, then the 
\f4NL_TEXTMAX\f1 define is used for the message text length.  If the 
\f4\-t\f1 option parameter is present, then the \f2textmax\f1 input must 
be a decimal number in the range of (0\-32767).  If \f2textmax\f1 is 
specified and is not in the range (0\-32767) or is not greater than 
\f4NL_TEXTMAX\f1, then the message text length maximum is set to 
\f4NL_TEXTMAX\f1.
.PP
The format of a message text source file is defined as follows.
Note that the fields of a message text source line are separated 
by a single ASCII space or tab character.
Any other ASCII spaces or tabs are considered as being part
of the subsequent field.
.TP 5
\f4$set  \f2n comment\f1
.br
Where \f2n\f1 specifies the set identifier of the 
following messages until the next 
\f4$set\f1,
\f4$delset\f1
or end-of-file appears.
\f2n\f1 must be  a number in the range (1\-{\f4NL_SETMAX\f1}).
Set identifiers 
within a single source file need not be contiguous.
Any string following the set identifier is treated as a comment.
If no
\f4$set\f1
directive is specified in a message text source file, all messages
will be located in 
the default message set
\f4NL_SETD\f1.
.P
.TP 5
\f4$delset  \f2n comment\f1
.br
Deletes message set 
\f2n\f1
from an 
existing message catalogue.
Any string following the set number is treated as a comment.
.sp .5
(Note:  if \f2n\f1 is not a valid set, it is ignored.)
.P
.TP 5
\f4$ \f2comment\f1
.br
A line beginning with a dollar symbol (\f4$\fP) followed by an ASCII space or
tab character is treated as a comment.
.TP 5
\f2m message text\f1
.br
The 
\f2m\f1
denotes the message identifier, 
which is a number in the range (1-{\f4NL_MSGMAX\f1}).
The
message text
is stored in the message catalogue 
with the set identifier 
specified by the last
\f4$set \f1
directive, and with message identifier
\f2m\f1.
If the 
message text
is empty, 
and an ASCII space or tab field separator is present, an empty string is
stored in the message catalogue.
If a message source line has a
message number, but neither a field separator nor 
message text,
the existing message with that number (if any) is deleted from the
catalogue.
Message identifiers 
need not be contiguous.
The length of
message text 
must be in the range (0\-{\f4NL_TEXTMAX\f1}).
.P
.TP 5
\f4$quote \f2c\f1
.br
This line specifies an optional quote character
\f2c\f1,
which can be used to surround
message text 
so that trailing spaces or null (empty) messages are visible in a 
message source line.
By default, or if an empty
\f4$quote\f1
directive is supplied, no quoting of
message text 
will be recognized.
.P
Empty lines in a message text source file are ignored.
.P
.br
.ne 15
Text strings can contain the special characters and 
escape sequences defined in the following table:
.TS
center box;
a a a
a ap-1 a.
Description	Symbol	Sequence
_
newline	NL(LF)	\en
horizontal tab	HT	\et
vertical tab	VT	\ev
backspace	BS	\eb
carriage return	CR	\er
form feed	FF	\ef
backslash	\e	\e\e
bit pattern	ddd	\eddd
.TE
.P
If the character following a backslash is not one of those specified, 
the backslash is ignored.
The escape sequence 
\f4\eddd\f1
consists of backslash followed by 1, 2, or 3 octal digits, 
which are taken to specify the value of the desired character.
.P
Backslash followed by an ASCII newline character is also used 
to continue a string on the following line.
Thus, the following two lines describe a single message string:
.br
.P
.DS I
.ft 4
.nf
    1 This line continues \e
    to the next line
.fi
.ft 1
.DE
.P
which is equivalent to:
.P
.br
.DS I
.ft 4
.nf
    1 This line continues to the next line
.fi
.ft 1
.DE
.SH FILES
.TP
\f4/usr/lib/locale/\f2locale\f4/LC_MESSAGES/uxmesg\f1
language-specific message file [See \f4LANG\fP on \f4environ\f1(5).]
.SH "SEE ALSO"
\f4mkmsgs\f1(1),
\f4catopen\f1(3C),
\f4catgets\f1(3C),
\f4catclose\f1(3C),
\f4gettxt\f1(3C),
\f4nl_types\f1(5)
.SH NOTES
This version of \f4gencat\fP is built on the
\f4mkmsgs\f1 utility.
The \f4gencat\fP database
consists of two files, \f2catfile\f4.m\f1, which is a 
\f4mkmsgs\f1 format catalogue, and \f2catfile\f1,
which contains the information required to translate
a set and message number into a simple message number that can be used
in a call to \f4gettxt\fP.
.PP
Using \f4gettxt\f1 constrains the catalogues to be located in a subdirectory
under \f4/usr/lib/locale\f1.
This restriction is lifted by placing only a symbolic
link to the catalogue in the directory \f4/usr/lib/locale/Xopen/LC_MESSAGES\fP
when the catalogue is opened.
It is this link that \fPgettxt\f1 uses when 
attempting to access the catalogue.
The link is removed when the catalogue is closed
but occasionally as applications exit abnormally without closing catalogues
redundant symbolic links will be left in the directory.
.PP
For compatibility with previous versions of \f4gencat\fP released in a number
of specialized internationalization products, the \f4\-m\f1 option is supplied.
This option will cause \f4gencat\fP to build a single file 
.I catfile 
which is compatible
with the format catalogues produced by the earlier versions.
The retrieval
routines detect the type of catalogue they are using and will act
appropriately.


eoe/cmd/gencat/gencat.c
/*	Copyright (c) 1990, 1991 UNIX System Laboratories, Inc.	*/
/*	Copyright (c) 1984, 1986, 1987, 1988, 1989, 1990 AT&T	*/
/*	  All Rights Reserved  	*/

/*	THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF     	*/
/*	UNIX System Laboratories, Inc.                     	*/
/*	The copyright notice above does not evidence any   	*/
/*	actual or intended publication of such source code.	*/

/*	Copyright (c) 1984, 1986, 1987, 1988, 1989 AT&T	*/
/*	  All Rights Reserved  	*/

/*	THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF AT&T	*/
/*	The copyright notice above does not evidence any   	*/
/*	actual or intended publication of such source code.	*/

#ident	"@(#)gencat:gencat.c	1.2.3.1"
#include <dirent.h>
#include <stdio.h>
#include <ctype.h>
#include <nl_types.h>
#include <locale.h>
#include <pfmt.h>
#include <errno.h>
#include <string.h>

/*
 * gencat : take a message source file and produce a message catalog.
 * usage : gencat [-lm] catfile [msgfile]
 * if msgfile is not specified, read from standard input.
 * Several message files can be specified
 */
char *msg_buf = NULL;		/* was msg_buf[NL_TEXTMAX] */
struct cat_set *sets = 0;
int curset = 1;
int is_set_1 = 1;
int no_output = 0;
int coff = 0;
int list = 0;
int malloc_format = 0;
int nl_textmax = NL_TEXTMAX;
int maxtextmax = 32768;
char catname[MAXNAMLEN];
static void copy(FILE *, FILE *);
void usage(int);
void cat_msg_build(char *, FILE *);
void cat_dump(char *);
void fatal(void);


extern int cat_build(char *);
extern void cat_mmp_dump(char *);

FILE *tempfile;
FILE *fd_temp;

const char
  nomem[] = ":12:Out of memory: %s\n",
  badread[] = ":13:Cannot read %s: %s\n",
  badreadtmp[] = ":14:Cannot read temporary file: %s\n",
  badwritetmp[] = ":15:Cannot write in temporary file: %s\n",
  badwrite[] = ":16:Cannot write %s: %s\n",
  badcreate[] = ":17:Cannot create %s: %s\n",
  badopen[] = ":18:Cannot open %s: %s\n",
  badseektmp[] = ":19:seek() failed in temporary file: %s\n",
  badsetnr[] = ":20:Invalid set number %d -- Ignored\n",
  badmsgnr[] = ":21:Invalid message number %d -- Ignored\n";

main(argc, argv)
  int argc;
  char **argv;
{
  FILE *fd;
  register int i;
  int cat_exists;
  int opt;
  extern int optind;
  char	*asciimax = NULL;

  (void)setlocale(LC_ALL, "");
  (void)setcat("uxmesg");
  (void)setlabel("UX:gencat");
  /*
   * No arguments at all : give usage
   */
  if (argc == 1)
    usage(1);
    
  /*
   * Create tempfile
   */
  if ((tempfile = tmpfile()) == NULL){
    pfmt(stderr, MM_ERROR, ":22:tempfile() failed: %s\n",
      strerror(errno));
    exit(1);
  }
  
  /*
   * test for mkmsgs format
   */
  while ((opt = getopt(argc, argv, "t:lm")) != EOF){
    switch (opt) {
      case 'l' :
	list = 1;
	break;
      case 'm' :
	malloc_format = 1;
	break;
      case 't' :
	asciimax = optarg;
	break;
      default :
	usage(0);
	exit(1);

    }
  }
      
  if (optind == argc)
    usage(1);

  if(asciimax) {
	int badnum = 0;
	int j, k;

	for(j=0, k = strlen(asciimax); j<k; j++) {
		if(isdigit((int)asciimax[j] == 0))
			badnum++;
	}
	if(!badnum) {
		int tmp = atoi(asciimax);

		if((tmp > 0) && (tmp < maxtextmax) && (tmp > nl_textmax))
			nl_textmax = tmp;
	}
  }
  if((msg_buf = (char *)malloc(nl_textmax)) == NULL) {
	pfmt(stderr, MM_ERROR, nomem, strerror(errno));
	fatal();
  }
  /*
   * Check catfile name : if file exists, read it.
   */
  strcpy(catname, argv[optind]);
  
  /*
   * use an existing catalog
   */
  if (cat_build(catname))
    cat_exists = 1;
  else
    cat_exists = 0;

  /*
   * Open msgfile(s) or use stdin and call handling proc
   */
  if (optind == argc - 1)
    cat_msg_build("stdin", stdin);
  else {
    for (i = optind + 1; i < argc ; i++){
      if ((fd = fopen(argv[i], "r")) == 0){
        /*
         * Cannot open file for reading
         */
        pfmt(stderr, MM_ERROR, badopen, argv[i], strerror(errno));
        continue;
      }
      if (argc - optind > 2)
        pfmt(stdout, MM_INFO, ":23:%s:\n", argv[i]);
      cat_msg_build(argv[i], fd);
      fclose(fd);
    }
  }

  if (!no_output){
    /*
     * Work is done. Time now to open catfile (for writing) and to dump
     * the result
     */

    if (malloc_format) 
	cat_dump(catname);
    else 
	cat_mmp_dump(catname);
  } else
    if (cat_exists)
      pfmt(stderr, MM_ERROR, ":24:%s not updated\n", catname);
    else
      pfmt(stderr, MM_ERROR, ":25:%s not created\n", catname);

  exit(0);
}

void
usage(complain)
int complain;
{
  if (complain)
    pfmt(stderr, MM_ERROR, ":26:Incorrect usage\n");
  pfmt(stderr, MM_ACTION, ":27:Usage: gencat [-m] catfile [msgfile ...]\n");
  exit(0);
}

char linebuf[BUFSIZ];

/*
 * Scan message file and complete the tables
 */
void
cat_msg_build(filename, fd)
  char *filename;
  FILE *fd;
{
  register int i, j;
  char quotechar = 0;
  char c;
  int msg_nr = 0;
  int msg_len;

  /*
   * always a set NL_SETD
   */
  curset=NL_SETD;
  add_set(curset);
  
  while (fgets(linebuf, BUFSIZ, fd) != 0){


    if ((i = skip_blanks(linebuf, 0)) == -1)
      continue;

    if (linebuf[i] == '$'){
      i += 1;
      /*
       * Handle commands or comments
       */
      if (strncmp(&linebuf[i], CMD_SET, CMD_SET_LEN) == 0){
        i += CMD_SET_LEN;

        /*
         * Change current set
         */
        if ((i = skip_blanks(linebuf, i)) == -1){
          pfmt(stderr, MM_WARNING, ":28:Incomplete set command -- Ignored\n");
          continue;
        }
        add_set(atoi(&linebuf[i]));
        continue;
      }
      if (strncmp(&linebuf[i], CMD_DELSET, CMD_DELSET_LEN) == 0){
  i += CMD_DELSET_LEN;

        /*
         * Delete named set
         */
        if ((i = skip_blanks(linebuf, i)) == -1){
          pfmt(stderr, MM_WARNING, ":29:Incomplete delset command -- Ignored\n")
          ;
          continue;
        }
        del_set(atoi(&linebuf[i]));
        continue;
      }
      if (strncmp(&linebuf[i], CMD_QUOTE, CMD_QUOTE_LEN) == 0){
        i += CMD_QUOTE_LEN;

        /*
         * Change quote character
         */
        if ((i = skip_blanks(linebuf, i)) == -1)
          quotechar = 0;
        else
          quotechar = linebuf[i];
        continue;
      }
      /*
       * Everything else is a comment
       */
      continue;
    }
    if (isdigit(linebuf[i])){
      msg_nr = 0;
      /*
       * A message line
       */
      while(isdigit(c = linebuf[i])){
        msg_nr *= 10;
        msg_nr += c - '0';
        i++;
      }
      j = i;
      if (linebuf[i] == '\n')
        del_msg(curset, msg_nr);
      else
      if ((i = skip_blanks(linebuf, i)) == -1)
	add_msg(curset,msg_nr,1,"");
      else {
        if ((msg_len = msg_conv(filename, curset, msg_nr, fd, linebuf + j,
        BUFSIZ, msg_buf, nl_textmax, quotechar)) == -1){
    no_output = 1;
          continue;
  }
        add_msg(curset, msg_nr, msg_len + 1, msg_buf);
      }
      continue;
    }
    
    /*
     * Everything else is unexpected
     */
    pfmt(stderr, MM_WARNING, ":30:Unexpected line -- Skipped\n");
  }
}

/*
 * Skip blanks in a line buffer
 */
skip_blanks(linebuf, i)
  char *linebuf;
  int i;
{
  while (linebuf[i] && isspace(linebuf[i]) && !iscntrl(linebuf[i]))
    i++;
  if (!linebuf[i] || linebuf[i] == '\n')
    return -1;
  return i;
}

/*
 * Dump the internal structure into the catfile.
 */
void
cat_dump(catalog)
char *catalog;
{
  FILE *fd;
  FILE *f_shdr, *f_mhdr, *f_msg;
  long o_shdr, o_mhdr, o_msg;
  struct cat_set *p_sets;
  struct cat_msg *p_msg;
  struct cat_hdr hdr;
  struct _cat_set_hdr shdr;
  struct _cat_msg_hdr mhdr;
  int msg_len;
  int nmsg = 0;
  
  if ((fd = fopen(catalog,"w")) == NULL) {
    pfmt(stderr, MM_ERROR, badcreate, catalog, strerror(errno));
    fatal();
  }
    
  if ((f_shdr = tmpfile()) == NULL || (f_mhdr = tmpfile()) == NULL ||
      (f_msg = tmpfile()) == NULL){
    pfmt(stderr, MM_ERROR, ":31:Cannot create temporary file: %s\n",
      strerror(errno));
    fatal();
  }
  o_shdr = 0;
  o_mhdr = 0;
  o_msg = 0;
  
  p_sets = sets;
  hdr.hdr_set_nr = 0;
  hdr.hdr_magic = CAT_MAGIC;
  while (p_sets != 0){
    hdr.hdr_set_nr++;
    p_msg = p_sets->set_msg;
    
    /*
     * Keep offset in shdr temp file to mark the set's begin
     */
    shdr.shdr_msg = nmsg;
    shdr.shdr_msg_nr = 0;
    shdr.shdr_set_nr = p_sets->set_nr;
    while (p_msg != 0){
      shdr.shdr_msg_nr++;
      nmsg++;
      msg_len = p_msg->msg_len;
      /*
       * Get message from main temp file
       */
      if (fseek(tempfile, p_msg->msg_off, 0) != 0){
        pfmt(stderr, MM_ERROR, badseektmp, strerror(errno));
        fatal();
      }
      if (fread(msg_buf, 1, msg_len, tempfile) != msg_len){
        pfmt(stderr, MM_ERROR, badreadtmp, strerror(errno));
        fatal();
      }
      
      /*
       * Put it in the messages temp file and keep offset
       */
      mhdr.msg_ptr = (int)ftell(f_msg);
      mhdr.msg_len = msg_len;
      mhdr.msg_nr  = p_msg->msg_nr;
      if (fwrite(msg_buf, 1, msg_len, f_msg) != msg_len){
        pfmt(stderr, MM_ERROR, badwritetmp, strerror(errno));
        fatal();
      }

      /*
       * Put message header
       */
      if (fwrite((char *)&mhdr, sizeof(mhdr), 1, f_mhdr) != 1){
        pfmt(stderr, MM_ERROR, badwritetmp, strerror(errno));
        ;
        exit(1);
      }
      p_msg = p_msg->msg_next;
    }
    
    /*
     * Put set hdr
     */
    if (fwrite((char *)&shdr, sizeof(shdr), 1, f_shdr) != 1){
      pfmt(stderr, MM_ERROR, badwritetmp, strerror(errno));
      fatal();
    }
    p_sets = p_sets->set_next;
  }
  
  /*
   * Fill file header
   */
  hdr.hdr_mem = ftell(f_shdr) + ftell(f_mhdr) + ftell(f_msg);
  hdr.hdr_off_msg_hdr = ftell(f_shdr);
  hdr.hdr_off_msg = hdr.hdr_off_msg_hdr + ftell(f_mhdr);

  /*
   * Generate catfile
   */
  if (fwrite((char *)&hdr, sizeof (hdr), 1, fd) != 1){
    pfmt(stderr, MM_ERROR, badwrite, catname, strerror(errno));
    fatal();
  }
  
  copy(fd, f_shdr);
  copy(fd, f_mhdr);
  copy(fd, f_msg);
  fclose(f_shdr);
  fclose(f_mhdr);
  fclose(f_msg);
}

static char copybuf[BUFSIZ];
static void
copy(dst, src)
  FILE *dst, *src;
{
  int n;
  
  rewind(src);
  
  while ((n = fread(copybuf, 1, BUFSIZ, src)) > 0){
    if (fwrite(copybuf, 1, n, dst) != n){
      pfmt(stderr, MM_ERROR, badwrite, catname, strerror(errno));
      fatal();
    }
  }
}

void
fatal(void)
{
  exit(1);
}
